_config/oembed.yml

```
@@ -3,4 +3,4 @@ Name: coreoembed
 ---
 SilverStripe\Core\Injector\Injector:
   SilverStripe\View\Embed\Embeddable:
-    class: SilverStripe\View\Embed\EmbedResource
+    class: SilverStripe\View\Embed\EmbedContainer
```

We're not using the Injector correctly in this PR. In the past, we've always tried to shift as much of the instantiation work to injector.
- `Embed\Embed` should be created from an injector definition and get it's reference to crawler directly from there.
- `Embed\Embed` should be injected directly in `EmbedContainer`.

```suggestion
    class: SilverStripe\View\Embed\EmbedContainer
    properties:
      embed: '%$Embed\Embed'
  Embed\Embed:
    constructor:
      - '%$Embed\Http\Crawler'
```

---

```
    properties:
      embed: '%$Embed\Embed'
```

This approach does not work.  This only works for public properties, but not private or protected

```
There were 5 errors:

1) SilverStripe\View\Tests\Shortcodes\EmbedShortcodeProviderTest::testYoutube
Error: Cannot access protected property SilverStripe\View\Embed\EmbedContainer::$embed

/var/www/vendor/silverstripe/framework/src/Core/Injector/Injector.php:809
/var/www/vendor/silverstripe/framework/src/Core/Injector/Injector.php:651
/var/www/vendor/silverstripe/framework/src/Core/Injector/Injector.php:620
/var/www/vendor/silverstripe/framework/src/Core/Injector/Injector.php:1002
/var/www/vendor/silverstripe/framework/src/Core/Injector/Injector.php:955
/var/www/vendor/silverstripe/framework/src/Core/Injector/Injector.php:1125
/var/www/vendor/silverstripe/framework/src/Core/Injector/Injector.php:1113
/var/www/vendor/silverstripe/framework/src/View/Shortcodes/EmbedShortcodeProvider.php:88
/var/www/vendor/silverstripe/framework/tests/php/View/Shortcodes/EmbedShortcodeProviderTest.php:28
/var/www/vendor/silverstripe/framework/tests/php/View/Shortcodes/EmbedShortcodeProviderTest.php:47
/var/www/vendor/phpunit/phpunit/phpunit:98
```

---

```
  Embed\Embed:
    constructor:
      - '%$Embed\Http\Crawler'
```
`Embed\Embed` is a third party requirement.  `Embed::create()` is not available.  We could possibly use some flavor of `Injector::inst()->get(Embed::class)`, but this is strange since there's no intention of swapping out Embed::class for something else

---

We've taken this approach in the past ... for example, in our [i18n config](https://github.com/silverstripe/silverstripe-framework/blob/4/_config/i18n.yml), we instantiate a bunch of Symfony class and pass them to each other.

When we don't use the injector to instantiate specific dependencies, we occasionally get people complaining that they can't override things in their project. 

Is this an approach we want to officially support? Probably not, but we don't need to foreclose it either. And when someone comes to us to complain that their obscure video service is not supported by `embed/embed`, we can point them to the Injector and tell them to subclass `embed/embed`.

---

This is just seems like over abstraction to solve a problem that does not exist.  At this point we only support EmbedContainer only anyway, so even Embeddable isn't really swappable

```php
        // Only EmbedContainer is currently supported
        if (!($embeddable instanceof EmbedContainer)) {
            throw new \RuntimeException('Emeddable must extend EmbedContainer');
        }
```

So making embed itself swapable just seems totally pointless and all it will do it is make the code more complex and harder to read.

---

As a further example, on EmbedContainer, `getEmbed(): Embed` returns an `Embed`.  So, right off the bat we'd need to wrap `Embed` in fairly pointless `EmbedInterace` interface and return that to get this to work.

And then whatever implements this new `EmbedInterace` would need to exactly match this method `$this->extractor = $this->getEmbed()->get($this->url);` - which is this https://github.com/oscarotero/Embed/blob/master/src/Embed.php#L21 AND return an `Extractor`.  

Trying to make Embed swappable at this point just seems ridiculous.  It also not the point of this whole exercise, which was to upgrade embed v3 to v4

```php
    /**
     * @return bool
     */
    public function validate()
    {
        return !empty($this->getExtractor()->code->html ?? '');
    }
    
    public function getExtractor(): Extractor
    {
        if (!$this->extractor) {
            $this->extractor = $this->getEmbed()->get($this->url);
        }
        return $this->extractor;
    }
    
    public function getEmbed(): Embed
    {
        if (!$this->embed) {
            $this->embed = new Embed($this->getCrawler());
        }
        return $this->embed;
    }
```

---

As a further example, on EmbedContainer, `getEmbed(): Embed` returns an `Embed`.  So, right off the bat we'd need to wrap `Embed` in fairly pointless `EmbedInterface` interface and return that to get this to work.

And then whatever implements this new `EmbedInterface` would need to exactly match this method `$this->extractor = $this->getEmbed()->get($this->url);` - which is this https://github.com/oscarotero/Embed/blob/master/src/Embed.php#L21 AND return an `Extractor`.  

Trying to make Embed swappable at this point just seems ridiculous.

It also not the point of this whole exercise, which was to upgrade embed v3 to v4

```php
    /**
     * @return bool
     */
    public function validate()
    {
        return !empty($this->getExtractor()->code->html ?? '');
    }
    
    public function getExtractor(): Extractor
    {
        if (!$this->extractor) {
            $this->extractor = $this->getEmbed()->get($this->url);
        }
        return $this->extractor;
    }
    
    public function getEmbed(): Embed
    {
        if (!$this->embed) {
            $this->embed = new Embed($this->getCrawler());
        }
        return $this->embed;
    }
```

---

Spoke offline about the need to still support injectable subclasses of Embed, so have updated accordingly